### ISSUE TITLE
DevTestLabs StopVM Task gets BadRequest instead of stopping VM

### ISSUE THREAD

**Issue by adrianimajion** [rfennell/AzurePipelines#466]
### Azure DevOps Extension you are using
- DevTestLabs
### Where are you running it?
- Azure DevOps Service (VSTS)
### Version of Extension/Task
- Current public 
### Expected behaviour and actual behaviour
Expected behaviour: The VM is stopped
Actual behaviour: The error "BadRequest : The request URL is not valid." is printed to the console 
### Steps to reproduce the problem
Create a new VM in DevTest Labs and use the StopVM task in a workflow to stop it once running. 

I tried entering the stop-azurermvm command for the same instance on Powershell myself while remoted into the VM and it worked, after I signed in through the oauth prompt. The StartVM task works fine. and I noticed that script prints some information like the environment and tenantid that doesn't show up when StopVM tries to run.

**Comment by adrianimajion** [comment]
I just noticed this issue was reported in the Q&A section on Visual Studio Marketplace as well, where rfennell requested a verbose debug log for the command. I've pasted the bottom of the log below, let me know if you'd like to see the entire thing. FWIW, I'm using this task in the gui editor. 

`##[debug][Common.Authentication]: Received token with LoginType 'OrgId', Tenant: '***', UserId: '***'
##[debug][Common.Authentication]: Checking token expiration, token expires '04/02/2019 20:06:12 +00:00' Comparing to '04/02/2019 19:06:17 +00:00' With threshold '00:05:00', calculated time until token expiry: '00:59:55.3027926'
##[debug]============================ HTTP REQUEST ============================
##[debug]
##[debug]HTTP Method:
##[debug]POST
##[debug]
##[debug]Absolute Uri:
##[debug]https://management.azure.com/subscriptions/cabc2835-a09a-4e22-96a6-6ffa658ee40d/resourceGroups/MyGroupName/providers/Microsoft.Compute/virtualMachines/MyVMName/Stop?api-version=2019-03-01
##[debug]
##[debug]Headers:
##[debug]User-Agent                    : AzurePowershell/v3.6.0.0,PSVersion/v5.0.10586.117
##[debug]ParameterSetName              : The resource Id.
##[debug]CommandName                   : Invoke-AzureRmResourceAction
##[debug]
##[debug]Body:
##[debug]
##[debug]
##[debug]
##[debug]============================ HTTP RESPONSE ============================
##[debug]
##[debug]Status Code:
##[debug]BadRequest
##[debug]
##[debug]Headers:
##[debug]Pragma                        : no-cache
##[debug]x-ms-ratelimit-remaining-resource: Microsoft.Compute/UpdateVM3Min;238,Microsoft.Compute/UpdateVM30Min;1198
##[debug]Strict-Transport-Security     : max-age=31536000; includeSubDomains
##[debug]x-ms-request-id               : 2dbe0087-815e-4ef4-90a7-14d086444c43
##[debug]Cache-Control                 : no-cache
##[debug]Server                        : Microsoft-HTTPAPI/2.0,Microsoft-HTTPAPI/2.0
##[debug]x-ms-ratelimit-remaining-subscription-writes: 1199
##[debug]x-ms-correlation-request-id   : 602c2100-b5e0-4c5d-a39e-d2232d691a53
##[debug]x-ms-routing-request-id       : CANADACENTRAL:20190402T190617Z:602c2100-b5e0-4c5d-a39e-d2232d691a53
##[debug]X-Content-Type-Options        : nosniff
##[debug]Date                          : Tue, 02 Apr 2019 19:06:17 GMT
##[debug]
##[debug]Body:
##[debug]{
##[debug]  "error": {
##[debug]    "code": "BadRequest",
##[debug]    "message": "The request URL is not valid."
##[debug]  }
##[debug]}
##[debug]
##[debug]
##[debug]Caught exception from task script.
##[debug]Error record:
##[debug]Handle-LastError : 
##[debug]BadRequest : The request URL is not valid.
##[debug]At D:\a\_tasks\DevTestLabsStartVm_aedfa360-b2a1-4dae-9c0d-19d6e9571949\1.1.1\StopVM.ps1:54 char:5
##[debug]+     Handle-LastError
##[debug]+     ~~~~~~~~~~~~~~~~
##[debug]    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorExcep    tion
##[debug]    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorExceptio    n,Handle-LastError
##[debug] 
##[debug]Script stack trace:
##[debug]at Handle-LastError, D:\a\_tasks\DevTestLabsStartVm_aedfa360-b2a1-4dae-9c0d-19d6e9571949\1.1.1\task-funcs.ps1: line 10
##[debug]at <ScriptBlock><trap>, D:\a\_tasks\DevTestLabsStartVm_aedfa360-b2a1-4dae-9c0d-19d6e9571949\1.1.1\StopVM.ps1: line 54
##[debug]at <ScriptBlock>, D:\a\_tasks\DevTestLabsStartVm_aedfa360-b2a1-4dae-9c0d-19d6e9571949\1.1.1\StopVM.ps1: line 93
##[debug]at <ScriptBlock>, <No file>: line 1
##[debug]at <ScriptBlock>, <No file>: line 22
##[debug]at <ScriptBlock>, <No file>: line 18
##[debug]at <ScriptBlock>, <No file>: line 1
##[debug]Exception:
##[debug]Microsoft.PowerShell.Commands.WriteErrorException: 
##[debug]BadRequest : The request URL is not valid.
##[error]
BadRequest : The request URL is not valid.
##[debug]Processed: ##vso[task.logissue type=error]%0ABadRequest : The request URL is not valid.
##[debug]Processed: ##vso[task.complete result=Failed]
##[section]Finishing: Stop a DevTest Labs VM`

**Comment by adrianimajion** [comment]
I just noticed this issue was reported in the Q&A section on Visual Studio Marketplace as well, where rfennell requested a verbose debug log for the command. I've pasted the bottom of the log below, let me know if you'd like to see the entire thing. FWIW, I'm using this task in the gui editor.



`
##[debug][Common.Authentication]: Received token with LoginType 'OrgId', Tenant: '***', UserId: '***'
##[debug][Common.Authentication]: Checking token expiration, token expires '04/02/2019 20:06:12 +00:00' Comparing to '04/02/2019 19:06:17 +00:00' With threshold '00:05:00', calculated time until token expiry: '00:59:55.3027926'
##[debug]============================ HTTP REQUEST ============================
##[debug]
##[debug]HTTP Method:
##[debug]POST
##[debug]
##[debug]Absolute Uri:
##[debug]https://management.azure.com/subscriptions/MySubscriptionId/resourceGroups/MyResourceGroup/providers/Microsoft.Compute/virtualMachines/MyVM/Stop?api-version=2019-03-01
##[debug]
##[debug]Headers:
##[debug]User-Agent                    : AzurePowershell/v3.6.0.0,PSVersion/v5.0.10586.117
##[debug]ParameterSetName              : The resource Id.
##[debug]CommandName                   : Invoke-AzureRmResourceAction
##[debug]
##[debug]Body:
##[debug]
##[debug]
##[debug]
##[debug]============================ HTTP RESPONSE ============================
##[debug]
##[debug]Status Code:
##[debug]BadRequest
##[debug]
##[debug]Headers:
##[debug]Pragma                        : no-cache
##[debug]x-ms-ratelimit-remaining-resource: Microsoft.Compute/UpdateVM3Min;238,Microsoft.Compute/UpdateVM30Min;1198
##[debug]Strict-Transport-Security     : max-age=31536000; includeSubDomains
##[debug]x-ms-request-id               : 2dbe0087-815e-4ef4-90a7-14d086444c43
##[debug]Cache-Control                 : no-cache
##[debug]Server                        : Microsoft-HTTPAPI/2.0,Microsoft-HTTPAPI/2.0
##[debug]x-ms-ratelimit-remaining-subscription-writes: 1199
##[debug]x-ms-correlation-request-id   : 602c2100-b5e0-4c5d-a39e-d2232d691a53
##[debug]x-ms-routing-request-id       : CANADACENTRAL:20190402T190617Z:602c2100-b5e0-4c5d-a39e-d2232d691a53
##[debug]X-Content-Type-Options        : nosniff
##[debug]Date                          : Tue, 02 Apr 2019 19:06:17 GMT
##[debug]
##[debug]Body:
##[debug]{
##[debug]  "error": {
##[debug]    "code": "BadRequest",
##[debug]    "message": "The request URL is not valid."
##[debug]  }
##[debug]}
##[debug]
##[debug]
##[debug]Caught exception from task script.
##[debug]Error record:
##[debug]Handle-LastError : 
##[debug]BadRequest : The request URL is not valid.
##[debug]At D:\a\_tasks\DevTestLabsStartVm_aedfa360-b2a1-4dae-9c0d-19d6e9571949\1.1.1\StopVM.ps1:54 char:5
##[debug]+     Handle-LastError
##[debug]+     ~~~~~~~~~~~~~~~~
##[debug]    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorExcep    tion
##[debug]    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorExceptio    n,Handle-LastError
##[debug] 
##[debug]Script stack trace:
##[debug]at Handle-LastError, D:\a\_tasks\DevTestLabsStartVm_aedfa360-b2a1-4dae-9c0d-19d6e9571949\1.1.1\task-funcs.ps1: line 10
##[debug]at <ScriptBlock><trap>, D:\a\_tasks\DevTestLabsStartVm_aedfa360-b2a1-4dae-9c0d-19d6e9571949\1.1.1\StopVM.ps1: line 54
##[debug]at <ScriptBlock>, D:\a\_tasks\DevTestLabsStartVm_aedfa360-b2a1-4dae-9c0d-19d6e9571949\1.1.1\StopVM.ps1: line 93
##[debug]at <ScriptBlock>, <No file>: line 1
##[debug]at <ScriptBlock>, <No file>: line 22
##[debug]at <ScriptBlock>, <No file>: line 18
##[debug]at <ScriptBlock>, <No file>: line 1
##[debug]Exception:
##[debug]Microsoft.PowerShell.Commands.WriteErrorException: 
##[debug]BadRequest : The request URL is not valid.
##[error]
BadRequest : The request URL is not valid.
##[debug]Processed: ##vso[task.logissue type=error]%0ABadRequest : The request URL is not valid.
##[debug]Processed: ##vso[task.complete result=Failed]
##[section]Finishing: Stop a DevTest Labs VM
`

**Comment by rfennell** [comment]
Does the start task work?

Remember the IDs it needs for the task are the Lab not the VM as per [the docs]((https://github.com/rfennell/AzurePipelines/wiki/DevTestLabs)

I have retested, see the run [here](https://dev.azure.com/richardfennell/GitHub/_releaseProgress?_a=release-pipeline-progress&releaseId=1296), and I can't get it to fail

**Comment by adrianimajion** [comment]
The start task does work - to work around the problem with the stop task, I'm running an Azure Powershell script to stop the VM, but I'm still using the start task. 

For the ID on both tasks I retrieved my VM's resource ID via powershell as directed by the docs. Just to confirm, this is the proper format, right? 
`/subscriptions/my-subscr-iption-id/resourceGroups/MYRESOURCEGROUP/providers/Microsoft.Compute/virtualMachines/virtualMachineName
`
I just tried again with a very simple test to make sure I hadn't missed something silly and encountered the same error behavior again. I created a pipeline with two agent jobs, one to start the VM and one to stop it, with the same exact resource ID in both tasks. The start VM task works, the stop VM task gets as far as "Stopping Lab VM 'name' from Lab ''" and produces the error. I noticed neither of these tasks are printing my lab name properly, but the start task doesn't otherwise fail. 

In the Powershell task I'm running `Stop-AzureRmVM -ResourceGroupName "MYRESOURCEGROUP" -Name "vmName" -Force` and that works. 

Definitely a tricky one, I know your tests are passing but I can't figure what I'm doing wrong or what could be misconfigured on my side. Let me know if there's a way I can help you diagnose, I'd be happy to do a screenshare session.

**Comment by adrianimajion** [comment]
Aha, I noticed the other individual asking about this in the marketplace Q&A and just found their closed issue, and you're right, I *am* calling the wrong ID, it just happens to work for the Start VM task as explained in that issue. I think it's worth adding an extra note to the documentation clarifying that the lab id is different from the resource id returned by get-azurermvm. I read the tooltip in the task for "Lab VM Id" and saw it described as "Resource ID of the lab VM" and promptly googled "get azure VM resource ID," which sent me towards the wrong commands. At least in my case, I already had created the lab VM I wanted to start and stop, so the output of the Create Azure DevTest Labs VM command was not readily available to me. What's the fastest way to get the Lab ID after I've already created the VM? For now I've just reformatted the resource id the way you indicated in the older issue. 

Ultimately still my bad reading comprehension, but I think the docs could be a bit more "user proof." Thanks for the help and for your work on these extensions, they make getting off the ground with Devops Pipelines much easier! And after this process I understand the Powershell side much better, so ultimately a very educational experience.
